### DISCUSSION TITLE
tRPC vs. react-query with manual types

### DISCUSSION THREAD

**Discussion by f0rr0** [trpc/trpc#2145] [Q&A] [answered]
Background:
- I understand how type safety works and why full-stack (end to end) type safety is great. 
- I understand how graphql provides this type safety but is not strongly integrated with typescript.
- I understand this project provides that stuff for typescript projects.
- I have played around with the idea of full-stack type safety.
- Worked with thrift and thrift compact protocol and code generation. Previous art: https://www.linkedin.com/pulse/type-safe-microservices-nodejs-thrift-typescript-kevin-greene/ , https://github.com/f0rr0/thrift-compact-protocol
- Worked with Protocol buffers client code generation.
- Worked with zod, react-query, Prisma.

Question:

- If I have an API route - `profile` and a handler defined at that route `getProfile` with a signature (let's use AWS lambda handler signature as an example):

```ts
export declare type PromiseHandler<TEvent = any, TResult = any> = (event: TEvent, context: Context) => Promise<TResult>;

// can be thought of as the core service contract
export interface UserProfileResponse {
  id: User['id']
  type: Entity.User
  data: Pick<User, 'name' | 'phone' | 'email' | 'address' | 'score'> & {
    inventory?: Record<string, boolean> | null
  }
}

let getProfile: PromiseHandler<
  APIGatewayProxyWithCognitoAuthorizerEvent,
  Response<UserProfileResponse>
> = (event) => {
  // handler code
}

export type GetUserProfileResponse = ThenArg<ReturnType<typeof getProfile>>
```
- On the client side, let's say I am using `react-query` like this project does internally and I consume the above API like so:

```ts
export const getUserProfile: QueryFunction<GetUserProfileResponse> = async () => {
  const { data } = await clientAxiosInstance.get<GetUserProfileResponse>(
    `/profile`
  )
  return data
}

// React function

const Profile = () => {
  const profileQuery = useQuery('profile', getUserProfile)
  
  // rest of the react function
}
```

- Let's say my typescript project is set to `strict` mode.

Is there anything this project will do (related to e2e type safety) that I don't have in the example above? If so what specifically are those things?

**Comment by mmkal** [comment, marked as answer]
Vs this example trpc would give you: input validation, less boilerplate on the server side, less boilerplate on the client side, cleaner and more readable code (IMHO), strongly typed procedure context, middleware, sensible/configurable request batching, a react-query caching strategy that works how you'd want it to without you needing to configure it, less tight coupling to AWS lambda/API gateway, and more type safety (you're relying on yourself opting in to the types for the `/profile` endpoint on the client side, where trpc would bind them automatically - and you're hardcoding the `/profile` string in the frontend and backend, then relying on good intentions to ensure they match up). 

But apart from that... not much?

Edit: I lied. There's plenty else. Your server would have an inspectable schema, and you'd have access to awesome community projects like trpc-playground, trpc-openapi, trpc-swr, loads more right now and whatever else comes in future.

And, (guessing here) unlike your hand-rolled kinda-strongly-typed solution, trpc has a pretty decent docs website which is super useful for new team members and lists and explains all of the above features. So it's actually a really good way of answering questions like

>Is there anything this project will do that I don't have in the example above? If so what specifically are those things?

**Reply by f0rr0** [reply]
Okay I think this is in line with what I expected. In a real world example, there would obviously be things like input validation, middleware and batching. Also I just used AWS lambda/API gateway as an example.

> you're relying on yourself opting in to the types for the /profile endpoint on the client side

Correct I could have easily written some other `XYZ` type there.

> where trpc would bind them automatically - and you're hardcoding the /profile string in the frontend and backend

ok this seems to be the only thing that the example misses from a guarantee standpoint (ignoring the beauty of code for a second and other features not related to e2e type safety). So whereas in my example above I as a human say if route is `profile` then type should be `GetUserProfileResponse`, this project uses its `router` semantics to register those things upfront against a given route.

I believe the router semantic should be a very small piece of code where we register the handler and client side query functions against a string route.

From the naming of this project and reading some of the intros it seems like the project has invented a new protocol or transport layer of some sort which guarantees type safety. From what I understood from your answer @mmkal - this project is a way to alleviate:
> where trpc would bind them automatically - and you're hardcoding the /profile string in the frontend and backend

The data is still serialised as json over the wire. At the client/server boundaries there is no explicit guarantee in the runtime that ensures type safety. Other than the feature mentioned above, there are some non related features (to e2e type safety) like middleware, batching, caching etc.

I concede that my example also does not guarantee the e2e type safety at runtime since we are still dealing with json and http / rest.
